src/unread/unreadModel.js

```diff
@@ -1,54 +1,187 @@
 /* @flow strict-local */
 import Immutable from 'immutable';
-import { createSelector } from 'reselect';
 
 import type { Action } from '../actionTypes';
 import type {
   UnreadState,
+  UnreadStreamsState,
   UnreadPmsState,
   UnreadHuddlesState,
   UnreadMentionsState,
 } from './unreadModelTypes';
 import type { GlobalState } from '../reduxTypes';
-import type { Selector } from '../types';
-import unreadStreamsReducer from './unreadStreamsReducer';
 import unreadPmsReducer from './unreadPmsReducer';
 import unreadHuddlesReducer from './unreadHuddlesReducer';
 import unreadMentionsReducer from './unreadMentionsReducer';
+import {
+  ACCOUNT_SWITCH,
+  EVENT_MESSAGE_DELETE,
+  EVENT_NEW_MESSAGE,
+  EVENT_UPDATE_MESSAGE_FLAGS,
+  LOGOUT,
+  MESSAGE_FETCH_COMPLETE,
+  REALM_INIT,
+} from '../actionConstants';
+import { getOwnUserId } from '../users/userSelectors';
 
-export const getUnreadStreams: Selector<
-  Immutable.Map<number, Immutable.Map<string, Immutable.List<number>>>,
-> = createSelector(
-  state => state.unread.streams,
-  data => {
-    // prettier-ignore
-    // This is an example of the style-guide rule "Always provide a type
-    // when writing an empty `Immutable` value".  Without the explicit type,
-    // `result` gets inferred as `Immutable.Map<number, empty>`, and then
-    // e.g. the `setIn` call could be completely wrong and the type-checker
-    // wouldn't notice.
-    const result: Immutable.Map<number, Immutable.Map<string, Immutable.List<number>>> =
-      Immutable.Map().asMutable();
-    for (const { stream_id, topic, unread_message_ids } of data) {
-      result.setIn([stream_id, topic], Immutable.List(unread_message_ids));
-    }
-    return result.asImmutable();
-  },
-);
+//
+//
+// Selectors.
+//
+
+export const getUnreadStreams = (state: GlobalState): UnreadStreamsState => state.unread.streams;
 
 export const getUnreadPms = (state: GlobalState): UnreadPmsState => state.unread.pms;
 
 export const getUnreadHuddles = (state: GlobalState): UnreadHuddlesState => state.unread.huddles;
 
 export const getUnreadMentions = (state: GlobalState): UnreadMentionsState => state.unread.mentions;
 
+//
+//
+// Reducer.
+//
+
+const initialStreamsState: UnreadStreamsState = Immutable.Map();
+
+// Like `Immutable.Map#map`, but with the update-only-if-different semantics
+// of `Immutable.Map#update`.  Kept for comparison to `updateAllAndPrune`.
+/* eslint-disable-next-line no-unused-vars */
+function updateAll<K, V>(map: Immutable.Map<K, V>, updater: V => V): Immutable.Map<K, V> {
+  return map.withMutations(mapMut => {
+    map.forEach((value, key) => {
+      const newValue = updater(value);
+      if (newValue !== value) {
+        mapMut.set(key, newValue);
+      }
+    });
+  });
+}
+
+// Like `updateAll`, but prune values equal to `zero` given by `updater`.
```

Comment:
Hmm true!

That never matters in our usage here, because we have an invariant that there never values equal to `zero`. But certainly the comment should match the behavior.

Effectively this lets the "update only if different" semantics of `update` take precedence over the "prune" semantics. I don't have a firm view on which way is better from an API-design perspective.